### ISSUE TITLE
Custom Moment type and ES2015 class syntax

### ISSUE THREAD

**Issue by darkadept** [jquense/yup#20]
I'm trying to grasp making my own type and I think I've got it now. I'm trying to use react-datepicker with react-formal. React-datepicker returns a Moment object and I'd like to keep it as a Moment object when I'm validating and casting using yup. I'm also using ES2015 class syntax, but it works using the old way too. This is what I've ended up with:

    class MomentSchema extends yup.mixed {
      constructor() {
        super({type: 'momentschema'});
        this.transforms.push(function(value) {
          if (this.isType(value)) return moment(value);
          return moment.invalid();
        });
      }
    }

From the yup docs it seems like I broke a cardinal sin by extending (inheriting) from yup.mixed. I tried it with the yup.date and it's fine but it returns JS dates only, not Moments.

I know I need to put more checks in this line: `if (this.isType(value)) return moment(value);` because it's accepting pretty much anything and running it through the moment constructor.

I use the above like this:

    const schema = yup.object({
      thedate: new MomentSchema(),
    });

(And to note, I had to wrap react-datepicker in a custom component that handles the state. But that's outside of the scope of this message.)

So am I on the right track here or am I misunderstanding the yup docs?

**Comment by jquense** [comment]
You are on the right track actually, the general rule is _only_ extend `yup.mixed` which is more of an abstract class, it gets a lot uglier when you try and extend anything of the other schema types. In fact they all inherit from mixed in the same way.

So in the constructor that initial transform will be your base parsing logic, the first thing that's run in a cast, so generally you first check if you need to do anything, (with `isType`) and if so coerce the value into the type you want.

```js
this.transforms.push(function(value) {
  // already a moment (or null if nullable() is set)
  if (this.isType(value)) return value;
  
  value =  moment(value);
  return value.isValid() ? value : moment.invalid()
});
```
depending on how you want to handle invalid dates.

What you are missing is that you need to implement a custom `_typeCheck` which will be used by the `isType()` method on your new schema. that's where you put your moment specific checks for whether the _type_ of value is correct.

You can see in the date schema that its:
```js
_typeCheck(v) {
  return isDate(v) && !isNaN(v.getTime())
},
```

which checks if its a date and a valid date, you'd probably want to assert that the value is a moment moment object.

**Comment by darkadept** [comment]
Excellent!

Yeah, I've been pouring over your source code to see how you do it for strings, date, etc. I thought I might need to doing something with the `_typeCheck()` function yet. And yeah, I definitely need to bolster the `transform()` function.

I actually only saw your datetime-picker just before I posted. I don't know how I missed it when reading over the React Formal docs. ;-)

It's dates (and moments) for now but I'm sure I'll have other custom data types that need special handling in the future. I found the documentation on creating custom types a bit hard to understand at first. The docs just inherit from the yup.date function which isn't really coming up with your own completely new custom type schema.

And thanks for the excellent libraries. After looking over many other React form and schema libraries (I'm coming from Meteor and SimpleSchema) yup and react-formal is simple yet extensible and fits my needs perfectly.
